### PR TITLE
Use pidfd to implement trio.Process.wait, when available

### LINKED ISSUE #1241: use pidfd on Linux when available
Recent versions of the Linux kernel finally added a much more reasonable way to interact with subprocesses called "pidfd". The main advantages are:

- Using an fd as a handle for interacting with a process avoids a bunch of weird edge cases that pids have, like pids being recycled out from under you so you accidentally kill the wrong process (this isn't an issue for the specific way trio interacts with subprocesses, because we only work with child processes and the relationship between a parent and child is special, but in more general cases it's an issue)
- You can wait for child processes using `epoll`!!!!
- There may be more fancy features coming that are built on top of pidfd, like the ability to automatically kill a child when the last pidfd reference to it is closed.

Probably the epoll integration is the first thing we care about, since it provides an alternative to the rube goldberg thread-based construction we use now. Of course we'll still have to maintain that indefinitely because of RHEL and friends, but it'll be nice to avoid it when running on a modern system. That also means this isn't like, a huge urgent priority, because the user-visible impact is basically nil. But we'll want it eventually, so might as well have an issue to collect information until we're ready.

Initial pidfd support landed in kernel 5.1; the epoll integration landed in 5.3. And Ubuntu 19.10 (out next week) will have a 5.3 kernel, so pidfd is starting to actually filter out to users. I believe that in 19.10 there still aren't any glibc wrappers for the new syscalls, so if you want to access them you have to [call `syscall` directly](https://manpages.courier-mta.org/htmlman2/syscall.2.html). We could do that with ctypes or something though – it's janky but not difficult.

The biggest blocker here is that we will definitely need to test the new code, and AFAIK none of the CI services we have access to will ship a pidfd-capable kernel for... a while. They all seem to stick to LTS releases, and sometimes they're slow to upgrade even then... Azure is still on Ubuntu 16.04, though Travis and Github Actions both have 18.04. So probably we won't be able to actually *test* pidfd support until sometime after Ubuntu 20.04 LTS is released in April 2020. (Unfortunately since this is a kernel feature, using docker to pull down a newer release doesn't help, and I'm not aware of any hosted CI services that let you bring-your-own-kernel.)

### PR DESCRIPTION
**Pull request by njsmith**
Fixes #1241

The new QEMU-based CI run (#1292) *should* let us test this. I'm curious
to see what codecov says :-).

But unfortunately, even with that, we're still only testing the
ctypes-based configuration, not the configuration that uses the stdlib
`os.pidfd_open` that currently only exists in CPython master. We're
running tests on CPython master, and we're running tests on a new
kernel, but we aren't doing both at the same time. AFAICT the only way
to test that configuration would be to check out and build CPython
master from inside the nested VM, which sounds like a real hassle.
OTOH, I'm not super comfortable about shipping code that hasn't been
tested... `os.pidfd_open` is pretty simple but it's always possible my
ctypes-based implementation has some subtle behavioral difference I
missed.

Also, maybe we should add some tests to confirm the pidfd is cleaned
up properly?

CC: @benjaminp

### CODE REVIEW

File: trio/_subprocess.py [hunk 1]
```diff
@@ -11,6 +12,39 @@
...
+        # fd and flags are actually int-sized, but the syscall() function
+        # always takes longs. (Except on x32 where long is 32-bits and syscall
+        # takes 64-bit arguments. But in the unlikely case that anyone is
+        # using x32, this will still work, b/c we only need to pass in 32 bits
+        # of data, and the C ABI doesn't distinguish between passing 32-bit vs
+        # 64-bit integers; our 32-bit values will get loaded into 64-bit
+        # registers where syscall() will find them.)
+        _cdll_for_pidfd_open.syscall.argtypes = [
+            ctypes.c_long,  # syscall number
+            ctypes.c_long,  # fd
```
Comment (oremanj) [inline trio/_subprocess.py]: I think this argument is called 'pid'?

File: trio/_subprocess.py [hunk 2]
```diff
@@ -214,10 +263,14 @@ def returncode(self):
         if self.poll() is None:
             async with self._wait_lock:
                 if self.poll() is None:
-                    await wait_child_exiting(self)
-                    self._proc.wait()
+                    if self._pidfd is not None:
+                        await trio.hazmat.wait_readable(self._pidfd)
+                    else:
+                        await wait_child_exiting(self)
+                    self.poll()
```
Comment (oremanj) [inline trio/_subprocess.py]: This is changing from `self._proc.wait()` to `self.poll()`. Any particular reason? I seem to recall that using `wait()` here was necessary for correctness on OS X, because kqueue and wait() aren't quite synchronized with each other, such that the waitpid(WNOHANG) used by poll() might return zero even though kqueue said "exiting" already.

Comment (njsmith) [reply trio/_subprocess.py]: Oof, good catch, I'd forgotten about that `waidpid(WNOHANG)` synchronization thing.

The idea was to centralize the logic for cleaning up the pidfd inside our `poll` method. But you're right, this needs a different approach.

File: trio/_subprocess.py [hunk 3]
```diff
@@ -31,7 +31,7 @@
         # registers where syscall() will find them.)
         _cdll_for_pidfd_open.syscall.argtypes = [
             ctypes.c_long,  # syscall number
-            ctypes.c_long,  # fd
+            ctypes.c_long,  # pid
```
Comment (pquentin) [inline trio/_subprocess.py]: nit: The comment above still refers to `fd`

**Comment by njsmith** [comment]
OK, I finally got back to this -- rebased on master, addressed the last review comment, I think it's ready to go.

**Comment by njsmith** [comment]
Build failures are #1478, so this is blocked on #1479 or similar.

**Comment by pquentin** [comment]
Now that #1479 is merged, is closing/reopening enough, or do we need to merge from master?

**Comment by njsmith** [comment]
I think closing/reopening should be enough, since the CI always does a merge with master before running. But let's try it and find out :-)

**Comment by pquentin** [comment]
GitHub won't let me merge because "Required statuses must pass before merging", but as far as I can tell everything passed. 🤔

**Comment by njsmith** [comment]
Yeah, I'm seeing that too. What the heck, github. I also manually compared the list of statuses here to the ones listed in the branch protection settings, and everything matches.
